### ISSUE TITLE
.NET 6:  Reusable Razor UI in class libraries with ASP.NET Core

### ISSUE THREAD

**Issue by rgamage** [dotnet/AspNetCore.Docs#24053]
In this article, there is an example shown to enable static web assets, but it is using the .NET 3.1/5 generic hosting model.  Need an example showing how to enable this using the minimal .NET 6 hosting model.

When the view tab is selected to show .NET 6.0, I would suggest the example should change from the old hosting model to the new model.

The existing .NET 5 example is shown like this:

```
    public static IHostBuilder CreateHostBuilder(string[] args) =>
        Host.CreateDefaultBuilder(args)
            .ConfigureWebHostDefaults(webBuilder =>
            {
                webBuilder.UseStaticWebAssets();
                webBuilder.UseStartup<Startup>();
            });
```

We need a similar example for the .NET 6 hosting model.  I have been unable to find any such example anywhere in the docs.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9fd019c7-ceb7-6ef3-26ba-6b44341b6fc5
* Version Independent ID: fd116d00-0dc7-efaf-c714-34d77b345a05
* Content: [Reusable Razor UI in class libraries with ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/razor-pages/ui-class?view=aspnetcore-6.0&tabs=visual-studio#consume-content-from-a-referenced-rcl)
* Content Source: [aspnetcore/razor-pages/ui-class.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/razor-pages/ui-class.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-razorpages**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

**Comment by Rick-Anderson** [comment]
We'll eventually get to this. It shouldn't be hard to covert to the new hosting model. See https://docs.microsoft.com/en-us/aspnet/core/migration/50-to-60-samples

**Comment by serpent5** [comment]
I had a quick look at this specific issue. Although it's possible to do this:

```csharp
builder.WebHost.UseStaticWebAssets();
```

That fails with an error about how the `WebRootPath` has changed. From what I can see, it isn't possible to do this with the minimal hosting model, but I could very well be wrong.

**Comment by Rick-Anderson** [comment]
@halter73 what's the right approach here for creating a Reusable Razor UI in class library with .NET 6?

**Comment by rgamage** [comment]
@Rick-Anderson your link to the 5.0 -> 6.0 examples does not show this feature, and this is why I raised this issue.  As @serpent5  mentioned, an exception is raised if you try the intuitive syntax shown.  I have a [stack overflow question](https://stackoverflow.com/questions/70083598/site-css-not-found-in-net-6-web-app-with-custom-environment) on this that has had no comments or answers.

**Comment by xantari** [comment]
Also running into this issue. Failure to set .UseStaticWebAssets() causes a lot of problems when your using build environment names such as "LocalDevelopment" instead of the "Development".

Which then of course breaks Razor's CSS Isolation feature as indicated here: https://github.com/dotnet/aspnetcore/issues/28174#issuecomment-734239932

In particular note this comment:

> Static web assets are only enabled by default in the Development environment. If you are using a custom environment you need to call webhost.UseStaticWebAssets() to enable them explicitly.

Which now means we can't enable them in .NET 6.0, cause there doesn't seem to be a way to call .UseStaticWebAssets anymore.

Which means it completely breaks CSS Isolation feature in this scenario.

If you go to your launchSettings.json file and put in the following:

```
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "LocalDevelopment"
      }
```

Then launch your project. You will notice that the following line in your _Layout.cshtml will no logner function:

```html
<link rel="stylesheet" href="~/My.Namespace.Web.styles.css" asp-append-version="true" />
```

Attempting to access that URL results in a 404. And no version string is appended to the link either.

Took me hours to find out that this appears to be an unfixable issue in .NET 6.0.

**Comment by xantari** [comment]
Actually it appears this is actually unfixable in .NET 6. This work around only half solves the problem:

```cs
//Put these two lines before builder.Build();
builder.WebHost.UseWebRoot("wwwroot");
builder.WebHost.UseStaticWebAssets();

var app = builder.Build();
```

While this does fix the <header> link tag to now generate the proper URL to the .css file and actually contains the CSS isolated classes. The rest of the page rendering is broken. Those CSS isolated changes are supposed to also roll down to the generated .cshtml content that the CSS isolation files belong to but they no longer do.

Example:

![image](https://user-images.githubusercontent.com/6610039/147386314-d505ec24-fd55-4d20-9d67-31e9da73adb0.png)

**Comment by rgamage** [comment]
@xantari - thanks for confirming, and for adding detailed documentation.  FYI there is a way to get this to work, but it's kind of a hacky work-around.  Answer is marked [here](https://stackoverflow.com/questions/70083598/site-css-not-found-in-net-6-web-app-with-custom-environment).  It would be nice if there were a way to do this correctly in .NET 6.

**Comment by brunolins16** [comment]
cc @dotnet/aspnet-blazor-eng

**Comment by javiercn** [comment]
@captainsafia can you recommend the approach here? (I believe you dealt with this on another issue?)

**Comment by captainsafia** [comment]
> That fails with an error about how the WebRootPath has changed. From what I can see, it isn't possible to do this with the minimal hosting model, but I could very well be wrong.

This particular issue should be fixed in 6.0.6 via https://github.com/dotnet/aspnetcore/pull/41182.

> Those CSS isolated changes are supposed to also roll down to the generated .cshtml content that the CSS isolation files belong to but they no longer do.

Your screenshot points that the unique attributes on the HTML elements is broken. Is there a mismatch between what is generated in the HTML and the CSS?

**Comment by javiercn** [comment]
> It really seems odd to me that we even have to do this for any environment. Why would this only be enabled in Dev and Prod environments for Razor CSS isolation to work? Can someone tell me why you would want to have it NOT work for all environments by default?

Because static web assets serves files from other locations like your nuget package cache in your machine and we don't want to be doing that unless we know we are in development. That's why if you are using a different environment you need to register it.

It is not needed in `Production` because at publish time the files get copied to the `wwwroot` folder and there is no runtime component at play.
